### PR TITLE
9297 feature flags

### LINKED ISSUE #9297: Create feature flags based on JVM settings to enable/disable experimental features
**Overview of the Feature Request**

For upcoming functionality like #9229 / #9268 / #8822 enabling and disabling this kind of features would benefit from a standardized mechanism based on `JvmSettings`. Usually this type of switch is called "feature flag" or "feature gate" and indicates something is early stage, experimental, opt-in behaviour.

**What kind of user is the feature intended for?**
Sysadmin

**What inspired the request?**
See issues above.

**What existing behavior do you want changed?**
Add additional configurability with flags under the scope of `dataverse.feature.[xxx]`. This scope shall be used for opting into using some experimental code or feature preview. Once a feature stabilizes and matures, the flag shall go away (maybe replacing with a dedicated configuration setting).

**Any brand new behavior do you want to add to Dataverse?**
It's arguable if we did not have feature flags before, as we have configuration mechanisms in place to enable/disable features. Yet, these configurations are usually meant to be for a stable, accepted functionality, while this "new" mechanism is dedicated to previews and experimentation.

**Any related open or closed issues to this feature request?**
#9229 / #9268 / #8822 and potentially more in the future.

### PR DESCRIPTION
**Pull request by poikilotherm**
**What this PR does / why we need it**:
Introduce feature flags to enable/disable _experimental_ and _previews_ of features.

**Which issue(s) this PR closes**:

Closes #9297

**Special notes for your reviewer**:
None so far.

**Suggestions on how to test this**:
Unit test is provided, otherwise nothing fancy to test here...? 

**Does this PR introduce a user interface change? If mockups are available, please link/include them here**:
Nope.

**Is there a release notes update needed for this change?**:
Not sure. Please provide guidance.

**Additional documentation**:
:battery: included.

### CODE REVIEW

File: doc/sphinx-guides/source/developers/configuration.rst [hunk 1]
```diff
@@ -109,3 +109,17 @@ always like ``dataverse.<scope/....>.newname...=old.property.name``. Note this d
 aliases.
 
 Details can be found in ``edu.harvard.iq.dataverse.settings.source.AliasConfigSource``
+
+Adding a Feature Flag
+^^^^^^^^^^^^^^^^^^^^^
+
+Some parts of our codebase might be opt-in only. Experimental or optional feature previews can be switched on using our
```
Comment (qqmyers) [inline doc/sphinx-guides/source/developers/configuration.rst:116]: Suggested addition after the first sentence. They may be only for developers/not intended for production, and/or temporary. (Experimental or optional features that can be used in production and are intended to remain, perhaps with some change over time, Experimental or optional feature previews...

Comment (poikilotherm) [reply doc/sphinx-guides/source/developers/configuration.rst:116]: Actually I think I got what causes the confusion here - we already do experimental things and simply used configuration and not flags for this before!

Here's a bold statement: as we seem to getting more and more of these now, how about we make all the other experimental things using this flag thing, too, so we are consistent again?

**Review by pdurbin — COMMENTED** [review]
I just looked at the code and the following docs:

- https://dataverse-guide--9299.org.readthedocs.build/en/9299/developers/configuration.html#adding-a-feature-flag
- https://dataverse-guide--9299.org.readthedocs.build/en/9299/installation/config.html#feature-flags

It all looks good to me. Can we move this to QA?

**Comment by poikilotherm** [comment]
@pdurbin any thoughts about renaming? Or making the scope smaller? Or migrating other experimental things to use this? (Could create a follow-up issue for this.)

File: doc/sphinx-guides/source/developers/configuration.rst (hunk 1 above)
Comment (qqmyers) [reply doc/sphinx-guides/source/developers/configuration.rst:116]: That would put #9290 which shouldn't get used in production and presumably will be going away back in the list of things you can use now and that we expect to stay around. I suppose that could be handled by strong messaging on that PR.

It also means anyone using these features has to reconfigure for 5.13? And we'd have to decide what to do with features that require some config info - do all their settings move to be under feature or do you turn them on one way and configure them elsewhere? Again, solvable problems, but I don't see a lot of benefit either.

This is definitely something where there probably isn't one right answer and I don't care much how it comes out as long as we don't confuse experimental things you can use in production with things that shouldn't through some form of messaging.

Comment (poikilotherm) [reply doc/sphinx-guides/source/developers/configuration.rst:116]: > That would put #9290 which shouldn't get used in production and presumably will be going away back in the list of things you can use now and that we expect to stay around. I suppose that could be handled by strong messaging on that PR.

That is a good and valid point - maybe there should be a forth column in the list that indicates security level? Maybe this should even be marked `@Deprecated` in code if it is serious for production so it becomes more visible this needs more work.

> It also means anyone using these features has to reconfigure for 5.13? And we'd have to decide what to do with features that require some config info - do all their settings move to be under feature or do you turn them on one way and configure them elsewhere? Again, solvable problems, but I don't see a lot of benefit either.

Well if we want to migrate existing things to this mechanism, we certainly will be able to make this backward compatible. (As done before for JvmSettings)
 
> This is definitely something where there probably isn't one right answer and I don't care much how it comes out as long as we don't confuse experimental things you can use in production with things that shouldn't through some form of messaging.

I agree. Another thought: while the session auth thing is maybe not perfectly secure, isn't this something that we probably will turn on during the process of moving to the SPA at least at demo.dataverse.org? We will need that session auth thing if we want to demonstrate using both in parallel. This kind of makes this thing a bit blury.

Comment (pdurbin) [reply doc/sphinx-guides/source/developers/configuration.rst:116]: I think the existing line conveys the main points:

- opt-in
- experimental
- optional

That said, no objection if someone wants to add more. To me, feature flags are a common term in the industry. Most developers have probably heard of them.

**Comment by poikilotherm** [comment]
Oh dear I think this was in ready by accident!!!! There was discussion ongoing but I had no time to address this yet!!!

**Comment by poikilotherm** [comment]
I wasn't aware that this is now a blocker for @GPortas - sorry! What I put in my old head was that you would do sth. to be non-blocked by this. Sorry for the misunderstanding involved!

I do have some stuff saved locally as a stash to evolve this into "EmergingFeature" and "ExperimentalFeature", addressing more of Jim's comments and which is an output of a discussion @GPortas and I had on Zoom. I don't recall if we took any notes. FWIW, I'll push this stuff to a branch as is so we can come back to it later if we need to.

As @pdurbin is happy with this PR and as it unblocks @GPortas, I'll put this back into review.

**Comment by pdurbin** [comment]
@poikilotherm thanks! ❤️ 

There are no new commits since I tested this, so I'll just go ahead and move it to QA.

**Review by pdurbin — APPROVED** [review]
I tested this and it seems to work well: https://github.com/IQSS/dataverse/pull/9299#issuecomment-1462518292

Moving to QA.

**Comment by mreekie** [comment]
grooming:
- This has been bumped back today from QA to read for qa
- however, we don't resize if it's made it to QA so no bleedover.

**Comment by pdurbin** [comment]
The API tests never ran. The spin up scripts died here:

```
TASK [dataverse : fire off installer] ******************************************
Connection to ec2-3-237-1-249.compute-1.amazonaws.com closed by remote host.
```

From https://jenkins.dataverse.org/job/IQSS-Dataverse-Develop-PR/job/PR-9299/8/console

**Comment by pdurbin** [comment]
I just merged the latest. Go, API tests! Go!

**Comment by pdurbin** [comment]
API tests passed (took about 10 minutes): https://jenkins.dataverse.org/job/IQSS-Dataverse-Develop-PR/job/PR-9299/9/testReport/edu.harvard.iq.dataverse.api/

**Pull request merged**